lib/std/reflection.nim

```diff
@@ -0,0 +1,80 @@
+##[
+This module implements some reflection API's.
+It is supported in c, cpp, js backends.
+
+Experimental API, subject to change.
+]##
+
+runnableExamples:
+  # `getProcname`
+  proc fn1(a: int, b = 'x') =
+    assert getProcname() == "fn1"
+    static: assert getProcname(withType = true) == "fn1: proc (a: int; b: char)"
+  fn1(1)
+
+  iterator fn3[T](a: T): string =
+    static: doAssert getProcname() == "fn3"
+    yield getProcname(withType = true)
+
+  from std/sequtils import toSeq
+  assert toSeq(fn3(1.5)) == @["fn3: proc (a: float64): string"]
+
+runnableExamples:
+  # `getExportcName`
+  from std/strutils import contains
+  proc fn1 =
+    assert "fn1" in getExportcName() # implementation defined, e.g. `fn1_reflection95examples49_1`
+  fn1()
+
+  proc fn2: auto {.exportc.} = getExportcName()
+  assert fn2() == "fn2"
+
+  iterator fn3[T](a: T): string {.closure.} =
+    yield getExportcName()
+
+  from std/sequtils import toSeq
+  let s1 = toSeq(fn3(1.5))[0]
+  let s2 = toSeq(fn3(1))[0]
+  assert "fn3" in s1
+  assert s2 != s1
+
+import std/macros
+
+template getExportcName*(): string =
+  ## Returns the name of the function containing the caller scope after codegen.
+  ## This is only valid inside a proc/func/method/closure iterator.
+  ## See also `getProcname`.
+  ##
+  ## .. note:: not supported in VM.
+  block:
+    var name {.inject.}: cstring
+    when defined(js):
+      {.emit: "`name` = arguments.callee.name;".}
+    else:
+      # C cast needed for cpp.
+      {.emit: "`name` = (char*)__func__;".}
+    $name
+
+macro getProcnameImpl(a: typed, withType: static bool): string =
+  let n = a.owner
+  var ret = n.repr
+  case n.symKind
+  of nskModule:
+    discard
```

Comment:
maybe raise

Reply:
i used to give a CT error here, but in practice, showing the module name instead was the more useful behavior, in particular for logging/debugging, so that code like this works:

```nim
template fn =
  echo getProcname() # works regardless if called from top-level or from inside a proc
  # do some work...
```

that said, i could renmae getProcname to `getOwnerName` which is more accurate

Reply:
I've now split in:
* getOwnerName (owner module, doens't care if it's a module or other symbol)
* getOwnerTypeString ("" if owner is module)

the API is cleaner this way.

getOwnerTypeString returning "" if owner is a module, for same reason i gave earlier: to avoid user having to special case when owner is module, when writing logging/debugging APIs